### PR TITLE
[ISSUE-661]Pipeline dependencies in multi-input ExecNode's decomposition

### LINKED ISSUE #661: Pipeline dependencies in multi-input ExecNode's decomposition.
## Enhancement

For a multi-input ExecNode, such as HashJoinNode, CrossJoinNode and etc.  left child is evaluated in ExecNode::open() at first, then chunks are pulled from right child one by one in ExecNode::get_next(). When a multi-input ExecNode(e.g. HashJoinNode) is decomposed into pipelines,  it is parted into two operators (HashJoinProbeOperator and HashJoinBuildOperator) for both sides. The operators corresponding left side and right side are denoted as left operator(HashJoinProbeOperator) and right operator(HashJoinBuildOperator)  respectively. The right operator(HashJoinBuildOperator) should be full materialized, so it is used as the SinkOperator of a pipeline, The left operator(HashJoinProbeOperator) can work in streaming-style, so it can appear in the middle of another pipeline. Denotes pipeline contains left operator and right operator as left pipeline and right pipeline  respectively, so left pipeline always has a data dependency on right pipeline, the left pipeline can be scheduled to execute on core until the left pipeline has finished. For a left pipeline, it shall get stuck in left operator.

OperatorWithDependency is introduced so that left operators should inherit from it. When a left pipeline get stuck
in left operator, it should be guarded by PipelineDriverPoller thread that will inspect std::vector<OperatorWithDependency*>
PipelineDriver::_dependencies to to ensure that left pipeline is put back into multi-level feedback queue after all left operators has been ready, i.e. OperatorWithDependency::is_ready() returns true.

### PR DESCRIPTION
**Pull request by satanson**
For a multi-input ExecNode, such as HashJoinNode, CrossJoinNode and etc. Right child is evaluated in ExecNode::open() at first, then chunks are pulled from left child one by one in ExecNode::get_next(). When a multi-input ExecNode(e.g. HashJoinNode) is decomposed into pipelines, it is parted into two operators (HashJoinProbeOperator and HashJoinBuildOperator) for both sides. The operators corresponding left side and right side are denoted as left operator(HashJoinProbeOperator) and right operator(HashJoinBuildOperator) respectively. The right operator(HashJoinBuildOperator) should be full materialized, so it is used as the SinkOperator of a pipeline, The left operator(HashJoinProbeOperator) can work in streaming-style, so it can appear in the middle of another pipeline. Denotes pipeline contains left operator and right operator as left pipeline and right pipeline respectively, so left pipeline always has a data dependency on right pipeline, the left pipeline can be scheduled to execute on core until the right pipeline has finished.

For a left pipeline, it shall get stuck in left operator. OperatorWithDependency is introduced so that left operators should inherit from it. When a left pipeline get stuck in left operator, it should be guarded by PipelineDriverPoller thread that will inspect std::vector<OperatorWithDependency*> PipelineDriver::_dependencies to to ensure that left pipeline can be put back into multi-level feedback queue after all right operators have been ready, i.e. OperatorWithDependency::is_ready() returns true.

### CODE REVIEW

File: be/src/exec/pipeline/operator_with_dependency.h [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+#include <string>
+#include <vector>
+
+#include "exec/pipeline/operator.h"
+#include "exec/pipeline/pipeline_fwd.h"
+
+namespace starrocks {
+namespace pipeline {
+// OperatorWithDependency is used to decompose multi-input ExecNode, such as HashJoinNode, CrossJoinNode and etc.
+// In multi-input ExecNode, left child is evaluated in ExecNode::open() at first, then chunks are pulled from
```
Comment (kangkaisen) [inline be/src/exec/pipeline/operator_with_dependency.h:13]:
```suggestion
// In multi-input ExecNode, right child is evaluated in ExecNode::open() at first, then chunks are pulled from
```

File: be/src/exec/pipeline/operator_with_dependency.h [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+// right child one by one in ExecNode::get_next(). Every multi-input ExecNode(e.g. HashJoinNode) is decomposed into
+// two operators (HashJoinProbeOperator and HashJoinBuildOperator) for both sides. The operators corresponding
+// left side and right side are denoted as left operator(HashJoinProbeOperator) and right operator(HashJoinBuildOperator)
+// respectively. The right operator(HashJoinBuildOperator) should be full materialized, so it is used as the SinkOperator
+// of a pipeline, The left operator(HashJoinProbeOperator) can work in streaming-style, so it can appear in the middle
+// of another pipeline. Denotes pipeline contains left operator and right operator as left pipeline and right pipeline
+// respectively, so left pipeline always has a data dependency on right pipeline, left pipeline can be scheduled to
+// execute on core until left pipeline has finished. For a left pipeline, it shall get stuck in left operator.
```
Comment (kangkaisen) [inline be/src/exec/pipeline/operator_with_dependency.h:21]:
```suggestion
// execute on core until right pipeline has finished. For a left pipeline, it shall get stuck in left operator.
```

File: be/src/exec/pipeline/operator_with_dependency.h [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+#include <vector>
+
+#include "exec/pipeline/operator.h"
+#include "exec/pipeline/pipeline_fwd.h"
+
+namespace starrocks {
+namespace pipeline {
+// OperatorWithDependency is used to decompose multi-input ExecNode, such as HashJoinNode, CrossJoinNode and etc.
+// In multi-input ExecNode, left child is evaluated in ExecNode::open() at first, then chunks are pulled from
+// right child one by one in ExecNode::get_next(). Every multi-input ExecNode(e.g. HashJoinNode) is decomposed into
```
Comment (kangkaisen) [inline be/src/exec/pipeline/operator_with_dependency.h:14]: left

File: be/src/exec/pipeline/pipeline_driver.h [hunk 4]
```diff
@@ -138,9 +142,20 @@ class PipelineDriver {
                _state == DriverState::INTERNAL_ERROR;
     }
     bool pending_finish() { return _state == DriverState::PENDING_FINISH; }
+    // return true if all the dependencies are ready, otherwise return false.
+    bool dependencies_stuck() {
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_driver.h:146]: Why not `dependencies_block`?

**Review by Pslydhh — DISMISSED** [review]

**Review by liuyehcf — DISMISSED** [review]

File: be/src/exec/pipeline/operator_with_dependency.h (hunk 1 above)
Comment (satanson) [reply be/src/exec/pipeline/operator_with_dependency.h:13]: fixed

File: be/src/exec/pipeline/operator_with_dependency.h (hunk 2 above)
Comment (satanson) [reply be/src/exec/pipeline/operator_with_dependency.h:21]: fixed

**Review by kangkaisen — APPROVED** [review]

**Review by Pslydhh — APPROVED** [review]

**Review by dirtysalt — APPROVED** [review]

**Pull request merged**